### PR TITLE
Gh 35 arg matcher

### LINKED ISSUE #35: Detect Arg matcher misuse
People often use `Arg.Any<int>()` in real calls (rather than `.Returns` stub or `.Received` assertion) to represent "I don't mind what input is used for this test".

This case is show in the [documentation](http://nsubstitute.github.io/help/argument-matchers/#how_not_to_use_argument_matchers). There is also an example in [this post](https://www.elasticmint.com/blog/2018/04/16/fixing-intermittent-test-failures/) (search for "Another example").

This causes problems for NSubstitute because it tries to use these argument matchers the next time it gets a call (it will consider the next call to be an attempt to stub, rather than a real call), and it can cause confusing test behaviour or test errors.

Proposed title:

> Argument matcher used without call to .Returns or .Received

Proposed description (from [docs](http://nsubstitute.github.io/help/argument-matchers/#how_not_to_use_argument_matchers)):

> Argument matchers should only be used when setting return values or checking received calls. Using Arg.Is or Arg.Any without a call to .Returns or Received() can cause your tests to behave in unexpected ways.

### PR DESCRIPTION
**Pull request by tpodolak**
I think this one is finally done. Would be nice if someone could test it before merging. I've tested it against known misusages and against the issues linked to the original task. As we don't have any private package registry I am uploading the package in here for testing

[NSubstitute.Analyzers.CSharp.1.0.11-beta-arg.zip](https://github.com/nsubstitute/NSubstitute.Analyzers/files/3614328/NSubstitute.Analyzers.CSharp.1.0.11-beta-arg.zip)

//cc @dtchepak @zvirja 

Closes #35

### CODE REVIEW

File: documentation/rules/NS5001.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+Or for a specific code block:
+````c#
+#pragma warning disable NS5001 // Argument matcher used without specifying a call.
+// the code which produces warning
+#pragma warning restore NS5001 // Argument matcher used without specifying a call.
+````
+
+## Additional information
+
+Unlike other analyzers this one produces warnings only during compilation of the project. This also means that when the problem is fixed by the user
```
Comment (dtchepak) [inline documentation/rules/NS5001.md]: This is a really helpful note. 👍

File: src/NSubstitute.Analyzers.VisualBasic/DiagnosticAnalyzers/ArgumentMatcherCompilationAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using NSubstitute.Analyzers.Shared;
+using NSubstitute.Analyzers.Shared.DiagnosticAnalyzers;
+
+namespace NSubstitute.Analyzers.VisualBasic.DiagnosticAnalyzers
+{
+    internal sealed class ArgumentMatcherCompilationAnalyzer : AbstractArgumentMatcherCompilationAnalyzer<InvocationExpressionSyntax, MemberAccessExpressionSyntax, ArgumentSyntax>
+    {
+        private static ImmutableArray<ImmutableArray<int>> AllowedPaths { get; } = ImmutableArray.Create(
```
Comment (dtchepak) [inline src/NSubstitute.Analyzers.VisualBasic/DiagnosticAnalyzers/ArgumentMatcherCompilationAnalyzer.cs]: These seem duplicated with the CSharp equivalent. Maybe move to base class and share the values?

Comment (tpodolak) [reply src/NSubstitute.Analyzers.VisualBasic/DiagnosticAnalyzers/ArgumentMatcherCompilationAnalyzer.cs]: Unfortunately, we cant. SyntaxKind is defined separately for VisualBasic and CSharp and the values of corresponding enums don't match. For instance
C#
```c#
ArgumentList = 8636, // 0x21BC
```
VB
```vb
ArgumentList = 347, // 0x015B
```

**Comment by dtchepak** [comment]
Found a minor edge case that mis-reports the error:

```csharp
namespace NSubWorkshop {
    using NSubstitute;
    using Xunit;

    public class SampleFixture {
        public interface ISample {
            void Example<T>(T o);
        }

        [Fact]
        public void Sample() {
            var sub = Substitute.For<ISample>();

            object arg = "hi";
            sub.Example(arg);

            sub.Received().Example((object)Arg.Is<string>(x => x == "hi"));
        }
    }
}
```

Casting the argument (normally done to ensure generic call matches) is detected as arg matcher misuse. I'm not sure how common this will be, so might be worth merging this as-is and addressing later?

**Comment by dtchepak** [comment]
Going through an older code-base I found it uses a particular pattern to auto-sub a result. Here is a fairly complete example:

```csharp
public interface IThing {
    string Name { get; set; }
}
public interface IThingLookup
{
    IThing GetThing(string name);
}
public interface IThingHelper
{
    void HelpThing(IThing thing);
}
public class SomeClass
{
    private readonly IThingLookup lookup;
    private readonly IThingHelper helper;

    public SomeClass(IThingLookup lookup, IThingHelper helper) {
        this.lookup = lookup;
        this.helper = helper;
    }

    public void DoStuff(string thingName) {
        helper.HelpThing(lookup.GetThing(thingName));
    }
}
[Fact]
public void SomeClassTest() {
    var lookup = Substitute.For<IThingLookup>();
    var helper = Substitute.For<IThingHelper>();

    var sut = new SomeClass(lookup, helper);

    // This will stub out an auto-value for `GetThing(Any string)` call:
    var thing = lookup.GetThing(Arg.Any<string>());
    // Equivalent to:
    //var thing = Substitute.For<IThing>();
    //lookup.GetThing(Arg.Any<string>()).Returns(thing);

    sut.DoStuff("any string");

    helper.Received().HelpThing(thing);
}
```

Rather than explicitly stubbing `lookup.GetThing(Arg.Any<string>())`, it uses NSubstitute's auto-subbing feature to save a line. (I don't like it, but it is used!)

An arg matcher is ok to pass to a substitute method, but not to a non-substitute (i.e. non-virtual, or on an instance that isn't a substitute like the [one from the docs](https://nsubstitute.github.io/help/argument-matchers/#how_not_to_use_argument_matchers), or on a data type (like the one listed as "another example" in [this post](https://www.elasticmint.com/blog/2018/04/16/fixing-intermittent-test-failures/)).

I guess another option is to detect the assignment. But having a line like:
```
lookup.GetThing(Arg.Any<string>());
```
is useless but not problematic unless `lookup` is not a substitute (or if `GetThing` is non-virtual).

**Comment by dtchepak** [comment]
Sorry @tpodolak, I have failed to consider that use case in [this comment](https://github.com/nsubstitute/NSubstitute.Analyzers/issues/35#issuecomment-417908969). 🤦‍♂ 😭  

It is a legitimate case that I've missed as I've tended to explicitly mock these cases.

**Comment by tpodolak** [comment]
> Found a minor edge case that mis-reports the error:
> 
> ```cs
> namespace NSubWorkshop {
>     using NSubstitute;
>     using Xunit;
> 
>     public class SampleFixture {
>         public interface ISample {
>             void Example<T>(T o);
>         }
> 
>         [Fact]
>         public void Sample() {
>             var sub = Substitute.For<ISample>();
> 
>             object arg = "hi";
>             sub.Example(arg);
> 
>             sub.Received().Example((object)Arg.Is<string>(x => x == "hi"));
>         }
>     }
> }
> ```
> 
> Casting the argument (normally done to ensure generic call matches) is detected as arg matcher misuse. I'm not sure how common this will be, so might be worth merging this as-is and addressing later?

Fix pushed




> Going through an older code-base I found it uses a particular pattern to auto-sub a result. Here is a fairly complete example:
> 
> ```cs
> public interface IThing {
>     string Name { get; set; }
> }
> public interface IThingLookup
> {
>     IThing GetThing(string name);
> }
> public interface IThingHelper
> {
>     void HelpThing(IThing thing);
> }
> public class SomeClass
> {
>     private readonly IThingLookup lookup;
>     private readonly IThingHelper helper;
> 
>     public SomeClass(IThingLookup lookup, IThingHelper helper) {
>         this.lookup = lookup;
>         this.helper = helper;
>     }
> 
>     public void DoStuff(string thingName) {
>         helper.HelpThing(lookup.GetThing(thingName));
>     }
> }
> [Fact]
> public void SomeClassTest() {
>     var lookup = Substitute.For<IThingLookup>();
>     var helper = Substitute.For<IThingHelper>();
> 
>     var sut = new SomeClass(lookup, helper);
> 
>     // This will stub out an auto-value for `GetThing(Any string)` call:
>     var thing = lookup.GetThing(Arg.Any<string>());
>     // Equivalent to:
>     //var thing = Substitute.For<IThing>();
>     //lookup.GetThing(Arg.Any<string>()).Returns(thing);
> 
>     sut.DoStuff("any string");
> 
>     helper.Received().HelpThing(thing);
> }
> ```
> 
> Rather than explicitly stubbing `lookup.GetThing(Arg.Any<string>())`, it uses NSubstitute's auto-subbing feature to save a line. (I don't like it, but it is used!)
> 
> An arg matcher is ok to pass to a substitute method, but not to a non-substitute (i.e. non-virtual, or on an instance that isn't a substitute like the [one from the docs](https://nsubstitute.github.io/help/argument-matchers/#how_not_to_use_argument_matchers), or on a data type (like the one listed as "another example" in [this post](https://www.elasticmint.com/blog/2018/04/16/fixing-intermittent-test-failures/)).
> 
> I guess another option is to detect the assignment. But having a line like:
> 
> ```
> lookup.GetThing(Arg.Any<string>());
> ```
> 
> is useless but not problematic unless `lookup` is not a substitute (or if `GetThing` is non-virtual).

If I understand you correctly, as this one
````c#
    var thing = lookup.GetThing(Arg.Any<string>());
````
is equvalent to 
````c#
    var thing = Substitute.For<IThing>();
    lookup.GetThing(Arg.Any<string>()).Returns(thing);
```` 
the whole problem of arg matcher misusage comes down to checking if Arg.Any/Arg.Is is used in non-virtual/non-interface method - just as in analyzers which check if Returns/Throws etc is used correctly?
Or there is a difference (from the NSubstitute perspective) between
````c#
var thing = lookup.GetThing(Arg.Any<string>());
````
and 
````c#
lookup.GetThing(Arg.Any<string>());
````
?

**Comment by dtchepak** [comment]
> the whole problem of arg matcher misusage comes down to checking if Arg.Any/Arg.Is is used in non-virtual/non-interface method - just as in analyzers which check usages of Returns/Throws etc?
> Or there is a difference (from the NSubstitute perspective) between
> 
> ```
> var thing = lookup.GetThing(Arg.Any<string>());
> ```
> and
> 
> ```
> lookup.GetThing(Arg.Any<string>());
> ```

I think that ends up being the main factor -- if they are used with a substitute and it is an overridable member then it should be ok. I don't think `lookup.GetThing(Arg.Any<string>());` will cause any problems for other tests. @zvirja can you please confirm this so I don't make more mistakes with this issue? 🤦‍♂

**Comment by dtchepak** [comment]
Hi @zvirja, if you get a chance could you please review the conversation in this thread? If you are too pressed for time please let us know and we'll press onwards. 😄

**Comment by zvirja** [comment]
Hi guys!

I was reviewing the https://github.com/nsubstitute/NSubstitute/issues/492 we have and the discussion about the same feature - usage of result of calls with matchers. And I found that we discussed the particularity of support, while we still agreed to support it.

> I don't think lookup.GetThing(Arg.Any<string>()); will cause any problems for other tests. 

I also think that _technically_ there shouldn't be issues. NSubsitute should handle those cases just fine - it will set call specification to context awaiting for further configuration actions, which might not come.

However, be aware that there also wrong usages likes this: https://github.com/nsubstitute/NSubstitute/issues/510, which also confuse people. Codewise they look the same and it's semantic which is different.

If you ask me, I never liked syntax like this:
```c#
sut.Echo(Arg.Is(str => str.StartsWith("Hello"))).Id.Returns("WelcomeId");
```
Instead, I think that we should use matchers only for configuration and never rely on values returned from such calls. But it's just my grumble opinion - I simply never used this feature 😅

**Comment by dtchepak** [comment]
> Instead, I think that we should use matchers only for configuration and never rely on values returned from such calls. But it's just my grumble opinion - I simply never used this feature

I prefer more explicit configuration too, but preference aside I think it is a valid use case, and I think it could cause trouble for Analyzers adoption if it is reported as an error.

I think the simplest thing that will catch most of the problematic cases is to ensure `Arg.xyz()` is passed to a substitutable member of a substitute (similar to `Returns` rules). Additional detection like ensuring the value is assigned I think adds complexity that I'm not sure is worthwhile for a severity of the cases it detects.

I think these cases are all ok:

```
var sub = Substitute.For< IThingLookup>();

// standard config
var thing1 = Substitute.For<IThing>();
sub.GetThing(Arg.Any<string>()).Returns(thing1);

// use auto-subbed value:
var thing2 = sub.GetThing(Arg.Any<string>());

// useless auto-subbed value:
sub.GetThing(Arg.Any<string>()); // returns unused auto-sub. Probably not useful, but shouldn't cause problems

// received:
sub.Received().GetThing(Arg.Any<string>());

//received in order:
Received.InOrder(() => { sub.GetThing(Arg.Is("hi")); sub.GetThing(Arg.Is("world")); } );

// arg do/invoke
sub.GetThing(Arg.Do<string>(x => ...));

// chained auto-subbed calls:
sub.GetThing(Arg.Any<string>()).Name.Returns("something");
```

If `GetThing` is non-substitutable (e.g. non-virtual) then these cases all become errors.

The current detection infrastructure in this PR looks amazing. I think we should consider keeping that for advanced rules and suggestions. For example transforming auto-sub configuration to explicit subs might be very useful:

```
var thing = sub.GetThing(Arg.Any<string>());
// transform to (or from):
var thing = Substitute.For<IThing>();
sub.GetThing(Arg.Any<string>()).Returns(thing);
```

**Comment by tpodolak** [comment]
@dtchepak I've updated analyzer and now it checks usages of Arg.Any and Arg.Is in non-virtual calls. There are couple of open points though:
- As discussed at the very beginning of this feature, analyzer was not supposed to check Arg.Do and Arg.Invoke usages. Is this still valid requirement, or should we extend the analyzer to look for those methods as well?
- As the logic now looks for usages in non-virtual calls, the [documentation](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/GH-35-arg-matcher/documentation/rules/NS5001.md) from this PR  is no longer valid. I can change it but then the explanation in the analyzers docs will be different than the one in the official [docs](https://nsubstitute.github.io/help/argument-matchers/#how_not_to_use_argument_matchers). Will we be aligning the official docs in this case or how should we proceed further?
- The analyzers which checks usages of Returns etc in non-virtual calls allows to suppress warnings in configuration file - see  [Safe API misuse](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/master/documentation/Configuration.md#safe-api-misuse) and https://github.com/nsubstitute/NSubstitute.Analyzers/issues/11. As the logic for arg matcher analyzer is similar and arg matchers are used in conjunction with Returns, should we also allow warning suppressions for specific methods for arg matcher rule?

**Comment by dtchepak** [comment]
Thanks @tpodolak! Apologies again for mis-defining this. :(

Let's look at the doc question first:

> As the logic now looks for usages in non-virtual calls, the documentation from this PR is no longer valid. I can change it but then the explanation in the analyzers docs will be different than the one in the official docs. Will we be aligning the official docs in this case or how should we proceed further?

I think the I will need to write new documentation for this. I wonder if this should be a new `NS1xxx` error, if it involves specifying a non-substitutable member? Could also maybe go in `NS3xxx` for arg spec problems, but I think the non-substitutable one is the main cause here. Let me know what `NS` code you think is best and I will try to write this doc this weekend and send a PR.

I think all the documentation for `NS5001` and the related official docs is still correct, it is just a warning rather than a definite error. 

 > analyzer was not supposed to check Arg.Do and Arg.Invoke usages. Is this still valid requirement, or should we extend the analyzer to look for those methods as well?

`Arg.Do` and `Arg.Invoke` should only be used with virtual members as well. This would put it in the new `NS1xxx` error code, but not in the old `NS5001` category.

> The analyzers which checks usages of Returns etc in non-virtual calls allows to suppress warnings in configuration file - see Safe API misuse and #11. As the logic for arg matcher analyzer is similar and arg matchers are used in conjunction with Returns, should we also allow warning suppressions for specific methods for arg matcher rule?

Yes I think suppressions of this would be good. 👍

**Comment by dtchepak** [comment]
@tpodolak Draft of documentation with identifier `NS1004` (happy to move category/number to whatever you prefer).

---

    # NS1004

    <table>
    <tr>
      <td>CheckId</td>
      <td>NS1004</td>
    </tr>
    <tr>
      <td>Category</td>
      <td>Non-substitutable member</td>
    </tr>
    </table>

    ## Cause

    Argument matcher used with a non-virtual member of a class.

    ## Rule description

    A violation of this rule occurs when an argument matcher such as `Arg.Is` or `Arg.Any` is used with a non-virtual member of a class. For example:

    ```c#
    // Incorrect if `NonVirtualMethod` is not virtual:
    sub.NonVirtualMethod(Arg.Any<int>()).Returns(42);
    ```

    This violation can also be reported if an argument matcher is used for something other than setting return values, checking received calls, or configuring callbacks. See [NSubstitute's argument matcher documentation](https://nsubstitute.github.io/help/argument-matchers/#how_not_to_use_argument_matchers) for more information.

    ## How to fix violations

    To fix a violation of this rule, make the member of your class virtual or substitute for interface.

    ## How to suppress violations

    This warning can be suppressed by disabling the warning in the **ruleset** file for the project.
    The warning can also be suppressed programmatically for an assembly:
    ````c#
    [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Non-substitutable member", "NS1004:Arg matcher with non-virtual member", Justification = "Reviewed")]
    ````

    Or for a specific code block:
    ````c#
    #pragma warning disable NS1004 // Arg matcher with non-virtual member.
    // the code which produces warning
    #pragma warning restore NS1004 // Arg matcher with non-virtual member.
    ````

**Comment by tpodolak** [comment]
@dtchepak NS1004 looks good to me - thanks for providing documentation. Analyzers were updated and now they are looking for usages of Arg matchers in non-virtual members.
When it comes to the second comment
> I think all the documentation for NS5001 and the related official docs is still correct, it is just a warning rather than a definite error.

Does it mean that we should keep NS5001 analyzer in place? Even though we are afraid that it might affect the adoption of analyzers?

> I prefer more explicit configuration too, but preference aside I think it is a valid use case, and I think it could cause trouble for Analyzers adoption if it is reported as an error.

Please note that analyzers don't report errors, (in our case)they always report warnings but quite a lot of people mark warnings as errors in the build. Analyzers can also report "information" diagnostic but these kinds of diagnostics are not visible in the build but only in the editor(with rather hard to spot icon) - but this is an option if we want to somehow show people missing explicit configuration

Visual Studio
image

Rider
image

**Comment by dtchepak** [comment]
Sorry for the delayed reply.

> Does it mean that we should keep NS5001 analyzer in place? Even though we are afraid that it might affect the adoption of analyzers?

I guess not. ☹️  I think NS5001 is valid advice and potentially useful information, but the risk of raising warnings for cases where people take advantage of understanding how the library works is probably too great. 

> Please note that analyzers don't report errors, (in our case)they always report warnings but quite a lot of people mark warnings as errors in the build. 

Ah ok, I always have warnings as errors so I often fail to make the distinction.

File: documentation/Configuration.md [hunk 3]
```diff
@@ -67,4 +67,4 @@ Defines target id for which given suppression is applicable. Target might be a n
 
 An array of rules to suppress.
 
-> :warning: As for today the only rule which can be safely suppressed is `NS1000`. See more on [GitHub](https://github.com/nsubstitute/NSubstitute.Analyzers/issues/11)
+> :warning: As for today the only rules which can be safely suppressed is `NS1000` and `NS1004`. See more on [GitHub](https://github.com/nsubstitute/NSubstitute.Analyzers/issues/11)
```
Comment (dtchepak) [inline documentation/Configuration.md]: Good catch 👍

**Pull request merged**